File: language/runtime/android/kotlin/src/main/kotlin/dev/destack/runtime/android/module/permission/RuntimeHostPermission.kt
package dev.destack.runtime.android.module.permission

import android.Manifest

import dev.destack.runtime.android.core.HostRequestId

/**
 * One Android permission selector.
 */
public enum class RuntimeHostPermission(
    /**
     * The stable ABI discriminant.
     */
    public val rawValue: Int,
) {
    /**
     * Location.
     */
    Location(1),

    /**
     * LocationBackground.
     */
    LocationBackground(2),

    /**
     * Camera.
     */
    Camera(3),

    /**
     * Microphone.
     */
    Microphone(4),

    /**
     * Bluetooth.
     */
    Bluetooth(5),

    /**
     * Notifications.
     */
    Notifications(6),

    /**
     * ContactsRead.
     */
    ContactsRead(7),

    /**
     * ContactsWrite.
     */
    ContactsWrite(8),

    /**
     * MediaRead.
     */
    MediaRead(9),

    /**
     * MediaWrite.
     */
    MediaWrite(10),

    /**
     * Motion.
     */
    Motion(11),

    /**
     * ClipboardRead.
     */
    ClipboardRead(12),

    /**
     * CalendarRead.
     */
    CalendarRead(13),

    /**
     * CalendarWrite.
     */
    CalendarWrite(14),
}

/**
 * Return the canonical host permission token for one runtime selector.
 */
internal fun runtimeHostPermissionName(
    permission: RuntimeHostPermission,
): String {
    return when (permission) {
        RuntimeHostPermission.Location -> "location"
        RuntimeHostPermission.LocationBackground -> "locationBackground"
        RuntimeHostPermission.Camera -> "camera"
        RuntimeHostPermission.Microphone -> "microphone"
        RuntimeHostPermission.Bluetooth -> "bluetooth"
        RuntimeHostPermission.Notifications -> "notifications"
        RuntimeHostPermission.ContactsRead -> "contactsRead"
        RuntimeHostPermission.ContactsWrite -> "contactsWrite"
        RuntimeHostPermission.MediaRead -> "mediaRead"
        RuntimeHostPermission.MediaWrite -> "mediaWrite"
        RuntimeHostPermission.Motion -> "motion"
        RuntimeHostPermission.ClipboardRead -> "clipboardRead"
        RuntimeHostPermission.CalendarRead -> "calendarRead"
        RuntimeHostPermission.CalendarWrite -> "calendarWrite"
    }
}

/**
 * Decode one runtime selector from one canonical host permission token.
 */
internal fun decodeRuntimeHostPermission(
    permissionName: String,
): RuntimeHostPermission? {
    return when (permissionName) {
        "location" -> RuntimeHostPermission.Location
        "locationBackground" -> RuntimeHostPermission.LocationBackground
        "camera" -> RuntimeHostPermission.Camera
        "microphone" -> RuntimeHostPermission.Microphone
        "bluetooth" -> RuntimeHostPermission.Bluetooth
        "notifications", "notification" -> RuntimeHostPermission.Notifications
        "contactsRead" -> RuntimeHostPermission.ContactsRead
        "contactsWrite" -> RuntimeHostPermission.ContactsWrite
        "mediaRead" -> RuntimeHostPermission.MediaRead
        "mediaWrite" -> RuntimeHostPermission.MediaWrite
        "motion" -> RuntimeHostPermission.Motion
        "clipboardRead" -> RuntimeHostPermission.ClipboardRead
        "calendarRead" -> RuntimeHostPermission.CalendarRead
        "calendarWrite" -> RuntimeHostPermission.CalendarWrite
        else -> null
    }
}

/**
 * One Android permission request submitted by one runtime session.
 */
public data class RuntimeHostPermissionRequest(
    /**
     * The stable request identifier for this interactive host flow.
     */
    val requestId: HostRequestId,

    /**
     * The normalized permission name requested by the runtime.
     */
    val permission: RuntimeHostPermission,
)

/**
 * One Android permission result event delivered into one runtime session.
 */
public data class RuntimeHostPermissionEvent(
    /**
     * The stable request identifier for this interactive host flow.
     */
    val requestId: HostRequestId,

    /**
     * The normalized permission name associated with this result.
     */
    val permission: RuntimeHostPermission,

    /**
     * Whether the Android host granted the permission.
     */
    val isGranted: Boolean,
)

/**
 * Return the Android permission constant for one runtime permission selector.
 */
internal fun androidPermissionName(
    permission: RuntimeHostPermission,
): String? {
    return when (permission) {
        RuntimeHostPermission.Location -> Manifest.permission.ACCESS_FINE_LOCATION
        RuntimeHostPermission.LocationBackground -> Manifest.permission.ACCESS_BACKGROUND_LOCATION
        RuntimeHostPermission.Camera -> Manifest.permission.CAMERA
        RuntimeHostPermission.Microphone -> Manifest.permission.RECORD_AUDIO
        RuntimeHostPermission.Bluetooth -> Manifest.permission.BLUETOOTH_CONNECT
        RuntimeHostPermission.ContactsRead -> Manifest.permission.READ_CONTACTS
        RuntimeHostPermission.ContactsWrite -> Manifest.permission.WRITE_CONTACTS
        RuntimeHostPermission.MediaRead -> Manifest.permission.READ_MEDIA_IMAGES
        RuntimeHostPermission.MediaWrite -> null
        RuntimeHostPermission.Motion -> Manifest.permission.ACTIVITY_RECOGNITION
        RuntimeHostPermission.CalendarRead -> Manifest.permission.READ_CALENDAR
        RuntimeHostPermission.CalendarWrite -> Manifest.permission.WRITE_CALENDAR
        RuntimeHostPermission.Notifications -> Manifest.permission.POST_NOTIFICATIONS
        RuntimeHostPermission.ClipboardRead -> null
    }
}

File: language/runtime/apple/core/Sources/RuntimeHostAppleCore/Module/Permission/RuntimeHostPermission.swift
import Foundation

/// One Apple permission selector.
public enum RuntimeHostPermission: Int32, Sendable, Codable {
  /// Location.
  case location = 1
  /// LocationBackground.
  case locationBackground = 2
  /// Camera.
  case camera = 3
  /// Microphone.
  case microphone = 4
  /// Bluetooth.
  case bluetooth = 5
  /// Notifications.
  case notifications = 6
  /// ContactsRead.
  case contactsRead = 7
  /// ContactsWrite.
  case contactsWrite = 8
  /// MediaRead.
  case mediaRead = 9
  /// MediaWrite.
  case mediaWrite = 10
  /// Motion.
  case motion = 11
  /// ClipboardRead.
  case clipboardRead = 12
  /// CalendarRead.
  case calendarRead = 13
  /// CalendarWrite.
  case calendarWrite = 14

  /// Return the canonical host permission token for this selector.
  public var hostName: String {
    switch self {
    case .location: "location"
    case .locationBackground: "locationBackground"
    case .camera: "camera"
    case .microphone: "microphone"
    case .bluetooth: "bluetooth"
    case .notifications: "notifications"
    case .contactsRead: "contactsRead"
    case .contactsWrite: "contactsWrite"
    case .mediaRead: "mediaRead"
    case .mediaWrite: "mediaWrite"
    case .motion: "motion"
    case .clipboardRead: "clipboardRead"
    case .calendarRead: "calendarRead"
    case .calendarWrite: "calendarWrite"
    }
  }

  /// Create one permission selector from one canonical host token.
  public init?(hostName: String) {
    switch hostName {
    case "location": self = .location
    case "locationBackground": self = .locationBackground
    case "camera": self = .camera
    case "microphone": self = .microphone
    case "bluetooth": self = .bluetooth
    case "notifications", "notification": self = .notifications
    case "contactsRead": self = .contactsRead
    case "contactsWrite": self = .contactsWrite
    case "mediaRead": self = .mediaRead
    case "mediaWrite": self = .mediaWrite
    case "motion": self = .motion
    case "clipboardRead": self = .clipboardRead
    case "calendarRead": self = .calendarRead
    case "calendarWrite": self = .calendarWrite
    default: return nil
    }
  }
}

/// One Apple permission request submitted by one runtime session.
public struct RuntimeHostPermissionRequest: Sendable, Hashable, Codable {
  /// The stable request identifier for this interactive host flow.
  public let requestID: HostRequestID
  /// The permission requested by the runtime.
  public let permission: RuntimeHostPermission

  /// Create one permission request.
  public init(
    requestID: HostRequestID,
    permission: RuntimeHostPermission
  ) {
    self.requestID = requestID
    self.permission = permission
  }

  /// Create one permission request from one canonical host token.
  public init(
    requestID: HostRequestID,
    permission: String
  ) {
    self.init(
      requestID: requestID,
      permission: requireRuntimeHostPermission(hostName: permission)
    )
  }
}

/// One Apple permission result event delivered into one runtime session.
public struct RuntimeHostPermissionEvent: Sendable, Hashable, Codable {
  /// The stable request identifier for this interactive host flow.
  public let requestID: HostRequestID
  /// The permission associated with this result.
  public let permission: RuntimeHostPermission
  /// Whether the Apple host granted the permission.
  public let isGranted: Bool

  /// Create one permission event.
  public init(
    requestID: HostRequestID,
    permission: RuntimeHostPermission,
    isGranted: Bool
  ) {
    self.requestID = requestID
    self.permission = permission
    self.isGranted = isGranted
  }

  /// Create one permission event from one canonical host token.
  public init(
    requestID: HostRequestID,
    permission: String,
    isGranted: Bool
  ) {
    self.init(
      requestID: requestID,
      permission: requireRuntimeHostPermission(hostName: permission),
      isGranted: isGranted
    )
  }
}

/// Compare one permission selector against one canonical host token.
public func == (left: RuntimeHostPermission, right: String) -> Bool {
  left.hostName == right
}

/// Compare one canonical host token against one permission selector.
public func == (left: String, right: RuntimeHostPermission) -> Bool {
  left == right.hostName
}

/// Resolve one permission selector from one canonical host token.
private func requireRuntimeHostPermission(hostName: String) -> RuntimeHostPermission {
  guard let permission = RuntimeHostPermission(hostName: hostName) else {
    preconditionFailure("invalid RuntimeHostPermission host token: \(hostName)")
  }

  return permission
}
